### ISSUE TITLE
Socket hang up after 2nd query

### ISSUE THREAD

**Issue by GGo3** [ClickHouse/clickhouse-js#150]
### Get this error after 2nd query to clickhouse
```
Error: socket hang up
    at connResetException (node:internal/errors:691:14)
    at Socket.socketOnEnd (node:_http_client:471:23)
    at Socket.emit (node:events:402:35)
    at endReadableNT (node:internal/streams/readable:1343:12)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  code: 'ECONNRESET'
}
```

### Steps to reproduce
1. Make 1 query with results
2. Make some actions with stat less than 7-10 sec
3. After that make one more query and get this error

Also, try to increase connect_timeout to 20 sec but it didn't help.

**Comment by slvrtrn** [comment]
What is the environment where the code is executed and the Node.js version?
Could you provide a minimal reproduction scenario?

**Comment by GGo3** [comment]
Sure
Node v16.13.2
@clickhouse/client 0.0.14
Clickhouse server 21.7.4.18

in createClient use only host username password database from .env
The problem is that when we have one open CH pool, we make select queries to the database one after another. When the interval between these queries is not long, up to 1-2 seconds, everything works smoothly. However, as soon as a longer interval is made between the queries (more then 5 sek), we receive this error. It feels like a feature was added where if the connection is not used for some time, the socket cannot be used.

**Comment by olexiyb** [comment]
We also try to understand the cause of ECONNRESET we see randomly
This article gives some [clue](https://connectreport.com/blog/tuning-http-keep-alive-in-node-js/)

**Comment by slvrtrn** [comment]
@GGo3, do you have a load balancer in front of ClickHouse? It could be one of the reasons for timeouts. Additionally, we just released 0.0.15 (props to @olexiyb); you could try checking it out, maybe the fix will work for you as well.

**Comment by GGo3** [comment]
Nope. Just remote clickhouse server, every 10 min by crontab start node script, createClient to CH and work with data.

**Comment by GGo3** [comment]
Updated new version 0.0.15. It doesn't help

**Comment by slvrtrn** [comment]
@GGo3, have you tried increasing `request_timeout`?

**Comment by slvrtrn** [comment]
I tried to reproduce it locally, but it worked fine.

My env: Fedora 37, Node.js v16.17.0, ClickHouse 23.3, @clickhouse/client 0.0.15

```js
import {createClient} from '@clickhouse/client'

void (async () => {
    const client = createClient({
        connect_timeout: 5_000,
        request_timeout: 60_000,
    })
    let rows
    for (let i = 1; i <= 10; i++) {
        rows = await longRunningQuery(client)
        console.info(`#${i}`, await rows.json())
        if (i !== 10) {
            await sleep(30_000)
        }
    }
    await client.close()
})()

function longRunningQuery(client) {
    return client.query({
        query: `SELECT sleep(3), rand64();`,
        format: 'JSONEachRow',
    })
}

function sleep(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
```

I suppose it is similar to the use case described in the OP.

It prints:

```
#1 [ { 'sleep(3)': 0, 'rand64()': '3559013576299350133' } ]
#2 [ { 'sleep(3)': 0, 'rand64()': '5520979173458390446' } ]
#3 [ { 'sleep(3)': 0, 'rand64()': '10411217679392783453' } ]
#4 [ { 'sleep(3)': 0, 'rand64()': '1204984556245591429' } ]
#5 [ { 'sleep(3)': 0, 'rand64()': '2927514430233340513' } ]
#6 [ { 'sleep(3)': 0, 'rand64()': '4869895186792984655' } ]
#7 [ { 'sleep(3)': 0, 'rand64()': '16062549944760496101' } ]
#8 [ { 'sleep(3)': 0, 'rand64()': '16442138677355003640' } ]
#9 [ { 'sleep(3)': 0, 'rand64()': '158837237593955734' } ]
#10 [ { 'sleep(3)': 0, 'rand64()': '7616247968259855513' } ]
```

and there are no exceptions.

Additionally, I tested it with our Cloud, similar code, only added `host` and `password`. Again, no exceptions.

**Comment by olexiyb** [comment]
Hi, my fix for timeout case was related to nodejs 19, I was able to fix it just because it was reproducible.
But we still see completely random `socket hang` up issue.
In our case we used Debian 11 + nodejs 18 + clickhouse 23.1 in production.
now we are migrating to Alpine 3.17 + nodejs 19 + clickhouse 23.3 and still see 1-2 hang issues per day
So far I was not able to identify the cause, but I see numerous reported issues in nodejs
https://github.com/nodejs/node/issues/39810
https://github.com/nodejs/node/issues/47228

Maybe we should use [undici](https://undici.nodejs.org/#/) for clickhouse client...

**Comment by nlapshin** [comment]
I see the same problem in my project. Ubuntu 22.04, nodejs 18.16 + clickhouse 22. I do not think that It depends of clickhouse or os version, more probably it depends of nodeJS version.

I changed versions node.js use 16 and 18 versions and caught socket hang up very often. My case: I call long http request to API(need to get big batch of data for one request) - it takes 1-2 minutes and after that try to delete(query below) old data and insert new batch of data. 

And I cannot see the same problem then run nodeJS 20 version.

`
ALTER TABLE {table: Identifier} DELETE where date between '2023-05-05' and '2023-05-05' "
`

**Comment by nlapshin** [comment]
And It does not depend that query I do. For example, I send

`
SELECT sleep(3), rand64();
`

For HTTP request I use node-fetch package, version 3.3.0

**Comment by Jank1310** [comment]
@olexiyb Did you find a fix to the hang ups / did you try undici?

**Comment by slvrtrn** [comment]
We reworked how request/socket timeouts work internally in [0.1.0](https://github.com/ClickHouse/clickhouse-js/releases/tag/0.1.0). Can someone please check if the issue is still reproducible after the upgrade?

**Comment by GGo3** [comment]
The problem is still presented. Update version from 0.0.14 to 0.1.0. Also, try to set 
connect_timeout: 5_000,
 request_timeout: 60_000
 
 Maybe need to test on the latest CH version, because we test it on 21.7.4.18

**Comment by GGo3** [comment]
Script run on server Ubuntu 18.04.4 LTS
Node v16.13.2
npm 8.1.2
@clickhouse/client 0.1.0

Clickhouse server 21.7.4.18

run script start with create an instance new Clickhouse(config).

CH:
const clickhouseClient = require("@clickhouse/client");

class Clickhouse {
    constructor (config) {
        if (!config.host || !config.user || !config.password || !config.database) {
            throw "Credentials not found";
        }

        const protocol = "http";
        const url = `${protocol}://${config.host}:${config.port || 8123}`;

        this.client = clickhouseClient.createClient({
            host: url,
            username: config.user,
            password: config.password,
            database: config.database,
            connect_timeout: 5_000,
            request_timeout: 60_000
        });
    }

    async query (sql) {
        const dataSet = await this.client.query({
            query: sql,
            format: "JSONEachRow"
        });

        const data = [];
        for await (const rows of dataSet.stream()) {
            for (let i = 0; i < rows.length; i++) {
                data.push(rows[i].json());
            }
        }

        return data;
    }

    async close () {
        await this.client.close();
    }
}

after that
Make 1 query with correct results
Make some actions with stat less than 7-10 sec
After that make 1 more query and get this error

**Comment by slvrtrn** [comment]
@GGo3, a few questions regarding the example:
- when `close` is called? 
- how the CH server is deployed? Any LB in front of it?

**Comment by GGo3** [comment]
1. Close only after the script is finished. I mean we wait for all queries result, it can be 10-30 query one by one.
2. CH server without LB on (Ubuntu 20.04.3 LTS)

**Comment by nlapshin** [comment]
@slvrtrn Today I updated to the latest version and It seems to work well. I'll try to get this into production next week and send my feedback later. Thanks.

**Comment by nlapshin** [comment]
@slvrtrn Today our team tested it and it seems the same problems are existed

**Comment by slvrtrn** [comment]
@GGo3, @nlapshin,
I managed to reproduce the issue locally.

Could you try tuning the [server keep alive setting](https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#keep-alive-timeout)?

For example, in `config.xml`:

```
<keep_alive_timeout>120</keep_alive_timeout>
```

The number here (in seconds) should be more than the expected idle of the application.

In the meantime, we are figuring out the proper solution on the client side.

**Comment by slvrtrn** [comment]
0.1.1 is out.

It introduces a new feature to track potentially expired sockets. 

Here's an excerpt from the docs:

-----

If you are experiencing `socket hang up` errors, there are several options to resolve this issue:

* Increase the value of the `keep_alive_timeout` server setting (`config.xml`), as it could be as little as 3s by default.
This could help if your application idles for slightly more than the default server setting. 
However, it is not always possible to increase it (for example, no access to the server's `config.xml`), 
and this setting shouldn't be increased to unreasonable values, 
and even then, a particular request can happen at an unfortunate time.
Expired socket detection feature can help in such situations.

* Enable expired socket detection and retry mechanism in the client:
```ts
const client = createClient({
  keep_alive: {
    enabled: true,
    // should be slightly less than the `keep_alive_timeout` setting in the server's `config.xml`
    // default is 3s there, so 2500 milliseconds seems to be a safe client value in this scenario
    // another example: if your configuration has `keep_alive_timeout` set to 60s, you could put 59_000 here
    socket_ttl: 2500,
    retry_on_expired_socket: true,
  },
})
```
If a potentially expired socket is detected (more than `socket_ttl` since that idle socket was used), 
and retry is enabled in the configuration, both socket and request will be immediately destroyed 
(before sending the data), and the client will recreate the request. 
Note that `socket_ttl` should be slightly less than the server `keep_alive_timeout` setting to make it work.
If `socket_ttl` is configured appropriately, it should reliably resolve `socket hang up` issues.

* As a last resort, it is possible to disable Keep-Alive feature entirely:
```ts
const client = createClient({
  keep_alive: {
    enabled: false,
  },
})
```

-----

@GGo3 @olexiyb @nlapshin 
could you test if `socket_ttl` + `retry_on_expired_socket` configuration will work for you?

**Comment by slvrtrn** [comment]
@movy, I agree that the KeepAlive topic is rather confusing, especially in Node.js with all the different timeouts and lack of proper socket housekeeping out of the box, etc.

Can you please share a snippet of your code that was derived from the endless stream example? Node.js/OS versions will also help.

I am curious how a hang-up error can happen when basically one connection is open indefinitely.

But the fact that `keep_alive: {enabled: false}` did not help as well is rather surprising. 

Usually, what happens is:
- With KeepAlive enabled, there is a pool of open sockets that HTTP Agent "manages".
- Since there is no reliable way to terminate idle sockets out of the box (even `agentkeepalive` library with `freeSocketTimeout` setting does not help), they are still considered "usable" even if the server has already closed the connection on its side
- when such an "expired" socket is assigned from the pool to a new request, and we try to send something, we have `socket hang up` cause the remote side is shut down.
- without KeepAlive, the sockets should be connected/reconnected every time we create a new request, so it should theoretically bypass the problem entirely at a cost of some performance.
 
But I do not expect this to be happening _without_ KeepAlive enabled.

Additionally, can you please confirm that this:

```ts
const client = createClient({
  keep_alive: {
    enabled: true,
    socket_ttl: 2500, // less than the server's KeepAlive timeout
    retry_on_expired_socket: true,
  },
})
```

does not help either? The entire idea here was to hack around the janky sockets management by trying to reliably detect an expiring one in advance. If even this does not help, we need to search for another solution (again).

@gabibora, please share the same info as requested above (Node.js version, OS, a minimal code snippet).

**Comment by movy** [comment]
Hello,
I'm running node v18.14.1 both on Ubuntu 22.04 and MacOS 13.3, relevant code:

```js
import { createClient } from '@clickhouse/client'
import Stream from 'node:stream'
import hyperid from 'hyperid'

export class ClickHouse {
    #client
    #streams = {}
    #insertPromises = {}
    #books

    constructor() {
        this.#client = createClient({
            host: process.env.CLICKHOUSE_HOST,
            username: 'default',
            password: process.env.CLICKHOUSE_SECRET,
            keep_alive: {
                enabled: true,
                socket_ttl: 2500,
                retry_on_expired_socket: true,
            },
        })

        for (const table of ['books', 'depth', 'mark_prices']) {
            this.#streams[table] = new Stream.Readable({
                objectMode: true,
                read() {},
            })
            this.#insertPromises[table] = this.#client
                .insert({
                    table,
                    values: this.#streams[table],
                    format: 'JSONEachRow',
                })
                .then(() => console.info(`${table} data ingestion is finished`))
                .catch(error => console.error)
        }
    }

    async init() {
        await this.#client.command({
            query: `
                    CREATE TABLE IF NOT EXISTS books (
                        id UUID,
                        time DateTime64,
                        symbol String,
                        askPrice Float64,
                        bidPrice Float64,
                    ) ENGINE = MergeTree()
                    PARTITION BY toYYYYMM(time)
                    ORDER BY (time, symbol)
                `
        }).catch(error => console.error())

        this.#books.emitter.on(`booksUpdate`, ({ symbol, bidPrice, askPrice }) => {
            this.#streams.books.push({ symbol, bidPrice: +bidPrice, askPrice: +askPrice, time: Date.now(), id: hyperid().uuid })
        })
    }

    // When app is closing, close the stream
    async closeStreams() {
        console.log('clickhouse cleanup')
        for (const stream of Object.values(this.#streams)) {
            stream.push(null)
        }
        // when the stream is closed, the insert stream can be awaited
        await Promise.all(Object.values(this.#insertPromises))
        await this.#client.close()
        console.log('clickhouse cleanup done')
    }
}

```

**Comment by slvrtrn** [comment]
@movy, thanks for sharing. For this particular use case, I'd just disable `keep_alive` as it does not add much here (aside from the issues) - all the sockets are actively used until the end of the application and not idling (except for one socket assigned to `command` during the init).

But if even disabled `keep_alive` causes issues... That is unexpected. I will set up a simple application derived from your example to do the background ingestion to see what happens in the long run vs CH Cloud.

In the Node.js docs, there is a proposed "workaround", which I was reluctant to include in the library core: https://nodejs.org/api/http.html#requestreusedsocket
They suggest just silently retrying in the event of a hanged socket if it's reused, which is what our [Python client actually does](https://github.com/ClickHouse/clickhouse-connect/blob/main/clickhouse_connect/driver/httpclient.py#L392-L410).

Regarding your code snippet and ECONNRESET: the client will not reconnect or do any retries (apart from when we detect an expired socket in advance using `retry_on_expired_socket`; if an error will pop up after the data is sent, it will just fail). I see that in the shared code, there is no recreation of a particular stream in case it's failed, only error logging. And you mentioned a dozen of such errors per day? Do you restart the application often? Cause effectively, I see that only 3 sockets are in use here with no retries (e.g. they will just be closed, and that's it).

One more question: how often are inserts triggered? is it a constant stream or with some idling sometimes?

**Comment by movy** [comment]
To clarify, by 'session' I mean running time between deploys. As I deploy (and restart the app) 2-3 times per day, I can see this error 2-3 times per "deploy" (consistently in the beginning, and sometimes later as well).

Streams are pretty constant (market data), couple seconds idling at most. 

I will disable `keep_alive` for now.
In your python source code I noticed this warning:
```py
                    raise ProgrammingError('Attempt to execute concurrent queries within the same session.' +
                                           'Please use a separate client instance per thread/process.')

```

Does it mean I have to instantiate separate client for each stream in NodeJS as well? Maybe it's a good idea regardless, did not cross my mind until now, I will add re-conneciton upon error as well, will see how it goes.

**Comment by slvrtrn** [comment]
@movy, that warning is for _multiprocessing_ scenario - Node.js is a single-threaded.

**Comment by movy** [comment]
It looks like with keep_alive off and client reconnection upon a error thrown, everything is pretty stable so far. I guess the docs need to mention keep_alive nuances and examples need to include proper error handling strategies. I wish I could help with that, but currently i'm en-route with very little spare time till August. If this still requires attention, I may PR then.

Thanks heaps for your help!

**Comment by slvrtrn** [comment]
@movy, how do you reconnect the client? By closing it (effectively destroying the underlying HTTP agent) and creating a new instance?

**Comment by amitava82** [comment]
Alright, I was going nuts with the same issue. Thankfully I found this thread. Is it recommended to keep the keep_alive off? I'd think it'll optimal to keep the connection open to reduce latency. In my APIs I'm getting lot of socket hang up error when user make frequent requests. Won't it add to latency with the TCP connection overhead?

Another question, I'm not calling `await client.close()` after ch query completed. Should I be doing it?

**Comment by slvrtrn** [comment]
@amitava82, you could try the workarounds described [in the docs](https://clickhouse.com/docs/en/integrations/language-clients/nodejs#keep-alive). 
Disabling KeepAlive will increase latency as the sockets with open connections won't be reused.
You should not call `await client.close()` after each query. Ideally, it should be called only once during the graceful shutdown of the app.

**Comment by amitava82** [comment]
Thanks! I'm trying out the suggestions. Not seeing any errors but I'll report if any.

**Comment by kobi-co** [comment]
@slvrtrn I have slightly different issue (with a similar error):
I have a clickhouse cluster we are connecting to using vpn. 
When I call `ping()` (When vpn is turned off) on this client instance, I'm getting a timeout error (which is OK), but in addition there in a uncaughtException which destroys my process: 
```
Error: socket hang up
    at connResetException (node:internal/errors:705:14)
    at TLSSocket.socketCloseListener (node:_http_client:467:25)
    at TLSSocket.emit (node:events:525:35)
    at TLSSocket.emit (node:domain:489:12)
    at node:net:301:12
    at TCP.done (node:_tls_wrap:588:7)
    at TCP.callbackTrampoline (node:internal/async_hooks:130:17) {
  code: 'ECONNRESET'
}
```

As opposed to this, When I use query in order to ping (tried a few different approaches), I don't get the same error (only timeout error without socket hang up that kills my process). 
Any idea here ?

Thanks!

**Comment by slvrtrn** [comment]
@kobi-co, this is interesting cause we explicitly destroy the response stream, and the entire call is wrapped in a try-catch in the internal implementation (see [this](https://github.com/ClickHouse/clickhouse-js/blob/main/packages/client-node/src/connection/node_base_connection.ts#L284-L299)).

Please confirm that you are running at least 0.2.1 and getting these errors there. I will try to debug it then.

**Comment by slvrtrn** [comment]
@kobi-co, yes, that's what I was thinking as well - the request stream is not destroyed properly. Thanks for the report!

**Comment by igolka97** [comment]
Getting same error so far with 2.3.0 ver. Also using streams as the participants above. Tried with or without keepalive option.

Env: node 18 alpine, CH 23.5.3.24. Everything running on docker separate containers.

**Comment by igolka97** [comment]
@slvrtrn 
- there are no any balancer between CH and application
- frequency is about 100-200 pushes to stream per second
- client config is simple as possible
```ts
return createClient({ host, username, password, database });
```

- sample code:

```ts
export class StreamRequestLoggerService<T extends PossibleLogRequest>
  extends RequestLoggerInterface<T>
  implements OnApplicationShutdown
{
  constructor(
    private readonly tableName: string,
    private readonly clickhouseClient: ClickHouseClient,
  ) {
    super();
    this.insertStream = this.clickhouseClient.insert({
      table: this.tableName,
      format: 'JSONEachRow',
      values: this.stream,
    });
  }

  stream = new Readable({
    objectMode: true,
    read: () => {
      /**/
    },
  });

  insertStream: Promise<ConnBaseResult>;

  onApplicationShutdown() {
    this.stream.push(null);
    return this.insertStream;
  }

  async logRequest(request: Array<T>) {
    request.forEach((r) => {
      this.stream.push(r);
    });
  }
}

**Comment by mshustov** [comment]
@igolka97 how do you use `logRequest`? Is the stream open all the time while the app running?

**Comment by igolka97** [comment]
@mshustov
yes, the stream is open all the time.
I'm using `logRequest` as `insert` abstraction over original  clickhouse-js `insert` method in my app. Calling `logRequest` each time as simple as that is when new http request is coming.

**Comment by slvrtrn** [comment]
@igolka97, the code you provided looks similar to what we have in our examples for endless streams, so I see no obvious issues.

I adjusted it slightly for my tests to specifically check the disabled keep alive setting, as you mentioned it:

```ts
import { ClickHouseClient, createClient } from "@clickhouse/client";
import { randomInt } from "crypto";
import { Readable } from "stream";

export class StreamRequestLoggerService {
  stream = new Readable({
    objectMode: true,
    read: () => {
      /**/
    },
  });

  insertStream: Promise<unknown>;

  constructor(
    private readonly tableName: string,
    private readonly clickhouseClient: ClickHouseClient,
  ) {
    this.insertStream = this.clickhouseClient
      .insert({
        table: this.tableName,
        format: "JSONEachRow",
        values: this.stream,
      })
      .then(() => console.info("\nData ingestion is finished"));
  }

  onApplicationShutdown() {
    this.stream.push(null);
    return this.insertStream;
  }

  logRequest<T>(request: Array<T>) {
    request.forEach((r) => {
      this.stream.push(r);
    });
  }
}

async function main() {
  const client = createClient({
    keep_alive: {
      enabled: false,
    },
  });
  const tableName = "test_logs";
  await client.command({
    query: `
      CREATE OR REPLACE TABLE ${tableName}
      (id UInt64, name String)
      ENGINE MergeTree()
      ORDER BY (id)
    `,
  });
  const log = new StreamRequestLoggerService("test_logs", client);

  for (let i = 0; i < 10_000; i++) {
    console.info(`[${i + 1}] Pushing several records into the stream...`);
    const data = [...Array(randomInt(100, 10_000))].map(() => ({
      id: randomInt(1, 100_000_000),
      name: Math.random().toString(36).slice(2),
    }));
    log.logRequest(data);
    await new Promise((resolve) => setTimeout(resolve, randomInt(1, 1000)));
  }

  // When Ctrl+C is pressed...
  async function cleanup() {
    await log.onApplicationShutdown();
    await client.close();
    process.exit(0);
  }

  process.on("SIGINT", cleanup);
  process.on("SIGTERM", cleanup);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
```

No issues with thousands of iterations here on my local setup...

The only way to trigger `ECONNRESET`/`socket hang up` was to kill the ClickHouse instance.

How often do you experience `socket hang up` in your case? Are you sure that the ClickHouse instance is not restarted at that time?

**Comment by movy** [comment]
I think I managed to create a reproducible test case, see below.

In my new project I use a universal Clickhouse class that is used for streaming inserts or for simple inserts / updates, depending on who's using the class. This lead me to discovery that if I create a `insertPromise` but never use it, normal inserts will fail within a minute after opening the connection (basically the opening case of this issue).

I realized this might be an anti-pattern in the realm of this library, so I tried creating separate clients for each function call (via `createClient(..)`), but this did not help. 

Eventually, I split my code into separate branches, so `insertPromise` is never created unless we actually intend to stream inserts on the instance.

After reading https://github.com/node-fetch/node-fetch/issues/1735 and https://github.com/vlucas/frisby/issues/594 I tried different versions of Node (17-21 I believe), same result everywhere. 

Please note that the error is not catchable, i.e. inevitably crashes the whole app each time it's encountered.  
Clickhouse 23.9.2.56, `"@clickhouse/client": "^0.2.4"`, Ubuntu 22.04

I tried myriad of `keep_alive` settings client and server-side, the results are basically the same. 

The code includes 4 cases, two of them will fail:

```js
import { createClient } from '@clickhouse/client'
import * as utils from './utils.js'
import Stream from 'node:stream'

export class ClickHouse {
    #insertPromise
    #client

    constructor() {
        this.#createStream()
        this.#client = createClient({
            host: process.env.CLICKHOUSE_HOST,
            username: 'default',
            password: process.env.CLICKHOUSE_SECRET,

            keep_alive: {
                enabled: false,
                // socket_ttl: 2500,
                // retry_on_expired_socket: true,
            }
        })
    }
    #createStream() {
        this.stream = new Stream.Readable({
            objectMode: true,
            read() {},
        })
    }
    async init() {
        await this.#client.command({
            // create test in memory table
            query: `
                CREATE TABLE IF NOT EXISTS test_table(
                    id UInt32,
                    name String,
                    time DateTime64
                ) ENGINE = Memory
                `,
        }).catch(error => console.error('⚠️ clickhouse init error:', error))
    }
    // await this.createPromise('videos_all')
    async createPromise(table) {
        this.#insertPromise = this.#client
            .insert({
                table,
                values: this.stream,
                format: 'JSONEachRow',
            })
            .catch(async (error) => {
                console.error(error)
                process.exit(255)
            })
    }

    async streamTestData() {
        console.log('streaming in test data')
        for (let index = 0; index < 1000; index++) {
            this.stream.push({
                id: index,
                name: 'test',
                time: Date.now(),
            })
        }
    }

    // insert data using INSERT (not stream) with 5 sec sleep in between
    async insertTestData() {
        for (let index = 0; index < 10; index++) {
            console.log('inserting test data', index)
            await this.#client.insert({
                table: 'test_table',
                values: [
                    {
                        id: index,
                        time: Date.now(),
                        name: 'test',
                    }
                ],
                format: 'JSONEachRow',
            }).catch(console.error)
            await utils.sleep(5000)
        }
    }
    async closeStreams() {
        // count rows in test_table
        const { data } = await (await this.#client.query({ query: 'SELECT count(*) FROM test_table' })).json()
        console.log('count rows in test_table', data)
        // close stream
        console.log('clickhouse cleanup')
        this.stream.push(null)
        // stream.destroy()
        // when the stream is closed, the insert stream can be awaited
        await this.#insertPromise
        // drop test table

        await this.#client.close()
        console.log('clickhouse cleanup done')
    }
}

// this passes
const test1 = async () => {
    const clickhouse = new ClickHouse()
    await clickhouse.init()
    // await clickhouse.createPromise('test_table')
    // await clickhouse.streamTestData()
    await clickhouse.insertTestData()
    await clickhouse.closeStreams()
}

// this passes
const test2 = async () => {
    const clickhouse = new ClickHouse()
    await clickhouse.init()
    await clickhouse.createPromise('test_table')
    await clickhouse.streamTestData()
    // await clickhouse.insertTestData()
    await clickhouse.closeStreams()
}

// this fails with SOCKET_TIMEOUT
const test3 = async () => {
    const clickhouse = new ClickHouse()
    await clickhouse.init()
    await clickhouse.createPromise('test_table')
    await clickhouse.streamTestData()
    await clickhouse.insertTestData()
    await clickhouse.closeStreams()
}

// this fails with Error: socket hang up ECONNRESET
const test4 = async () => {
    const clickhouse = new ClickHouse()
    await clickhouse.init()
    await clickhouse.createPromise('test_table')
    // await clickhouse.streamTestData()
    await clickhouse.insertTestData()
    await clickhouse.closeStreams()
}

// await test1()
// await test2()
// await test3()
await test4()

```

**Comment by slvrtrn** [comment]
@movy, I checked your example, but in my case (Node.js 18x., Fedora), all four pass. I see one socket hang up error at the very end when the application quits, though.

First take with the output (exact same code, just slightly adjusted to TS): https://gist.github.com/slvrtrn/392c6dd5371d84651a3c2d13ed32946a

Looks like socket hang up happens because we have some dangling promises produced here:

```ts
async createPromise(table) {
    this.#insertPromise = this.#client
        .insert({
            table,
            values: this.stream,
            format: 'JSONEachRow',
        })
        .catch(async (error) => {
            console.error(error)
            process.exit(255)
        })
}
```

And the stream is reused for all of these promises (which causes issues, cause we have some conflicting listeners now, i.e. UB).

Here is a fixed snippet (see some adjustments in `createPromise`): https://gist.github.com/slvrtrn/f920f042dea8e6ebfdb9728b124a5b71

NB: it also flushes the data correctly now, as it should be 2030 (and not 30) records there.

Enabling all KeepAlive settings with the fixed example like this:

```ts
keep_alive: {
  enabled: true,
  socket_ttl: 2500,
  retry_on_expired_socket: true,
}
```

causes no issues.

**Comment by movy** [comment]
@slvrtrn, I apologise for the confusion, but tests1-4 are to be run separately (hence I commented out all but one in my code). 

Thanks for your explanation regarding test4. Maybe it's worth to be documented, as may be obvious in hindsight, it might not be so for new library users. Basically it means we'd have to instantly push at least one record upon a promise creation, as in many practical use cases a promise is created on an app's launch, but data can come at any point later in time, not necessarily immediately after start. 3 seconds timeout is definitely insufficient here. 

One big problem with ECONNRESET is that it's uncatchable and often crashes the whole app without proper stack-trace, i.e. if we use many connections / streams and one of them hangs up like this, it is very tricky to find piece of code that caused it. 

As for test3, if run separately, the promise is utilised and the stream is destroyed correctly, yet still SOCKET_TIMEOUT error. I have never seen this in production though, only while creating this test-case.

**Comment by breslavsky** [comment]
I can confirm, we have the same issue:

```ts
return createClient({
    keep_alive: {
        enabled: true,
        socket_ttl: 60000,
        retry_on_expired_socket: true
    }
});
```

In Sentry 174 errors per 100k inserts to ClickHouse.

<img width="756" alt="Screenshot 2024-01-14 at 10 10 50" src="https://github.com/ClickHouse/clickhouse-js/assets/8783224/54d16fbf-232f-46e2-bc95-643ad6de5861">

**Comment by slvrtrn** [comment]
@breslavsky, why is your socket_ttl set to 60_000? what is your ClickHouse server `keep_alive_timeout` setting?

**Comment by breslavsky** [comment]
> @breslavsky, why is your `socket_ttl` set to 60_000? what is your ClickHouse server `keep_alive_timeout` setting?

We set 120 seconds.
Idea that client should close socket by itself every 60 seconds and it don't afraid that server unexpectedly closed the socket on their side.

**Comment by slvrtrn** [comment]
@breslavsky, could you DM me in the [community Slack](https://clickhouse.com/slack)?

**Comment by slvrtrn** [comment]
Closing this, as this should be fixed as of [0.3.0](https://github.com/ClickHouse/clickhouse-js/releases/tag/0.3.0). If there are still issues with sockets after upgrading, please open a new one.
